pkgs/build-support/fetchnodemodules/default.nix

```diff
@@ -0,0 +1,101 @@
+{ lib, stdenvNoCC, makeWrapper, nodejs }:
+
+{ src
+, hash ? ""
+, sha256 ? ""
+, name ? null
+, production ? true
+, runScripts ? false
+, makeTarball ? true
+, preferLocalBuild ? true
+, exposeBin ? false
+, npmFlags ? ""
+, ...
+} @ args:
+
+assert makeTarball -> !exposeBin;
+let
+  suffix = if makeTarball then ".tar.gz" else "";
+  name_ = if name == null then "${src.name}-node_modules${suffix}" else name;
+  nodejs_ = if args ? nodejs then args.nodejs else nodejs;
+
+  hash_ =
+    if hash != "" then { outputHashAlgo = null; outputHash = hash; }
+    else if sha256 != "" then { outputHashAlgo = "sha256"; outputHash = sha256; }
+    else throw "fetchNodeModules requires a hash for ${name_}";
+
+  customArgs = [ "name" "hash" "sha256" ];
+in stdenvNoCC.mkDerivation ({
+  inherit src production runScripts makeTarball preferLocalBuild;
+
+  name = name_;
+  nativeBuildInputs = [ nodejs_ makeWrapper ];
+
+  phases = "unpackPhase patchPhase buildPhase installPhase";
+
+  buildPhase = ''
+    if [[ ! -f package.json ]]; then
+        echo
+        echo "ERROR: The package.json file doesn't exist"
+        echo
+        exit 1
+    fi
+    if [[ ! -f package-lock.json ]]; then
+        echo
+        echo "ERROR: The package-lock.json file doesn't exist"
+        echo
+        echo "package-lock.json is required to make sure the node_modules"
+        echo "set we download never changes."
+        exit 1
+    fi
+
+    export npm_config_cache=/tmp
+    NPM_FLAGS="--no-update-notifier $npmFlags"
+
+    # Scripts may result in non-deterministic behavior.
+    # Some packages (e.g., Puppeteer) use postinstall scripts to download extra data.
+    if [[ -z $runScripts ]]; then
+        NPM_FLAGS+=" --ignore-scripts"
+    fi
+
+    # Whether to install dev dependencies
+    if [[ ! -z $production ]]; then
+        NPM_FLAGS+=" --production"
+    fi
+
+    echo "Running npm ci $NPM_FLAGS"
+    npm ci $NPM_FLAGS
+  '';
+
+  installPhase = ''
+    if [[ ! -z $makeTarball ]]; then
+        # Produce a deterministic tarball
+        tar -czf $out --owner=0 --group=0 --numeric-owner --format=gnu \
+            --mtime="@$SOURCE_DATE_EPOCH" --sort=name \
+            node_modules
+    else
+        # Copy all modules into a directory
+        mkdir $out
+        cp -r node_modules $out
```

Comment:
Could we move the files under lib, share or opt?

Reply:
Sure, it makes sense to follow the conventions since we are having a `bin` (if `exposeBin` is set) as well. Could you also test if you can reproduce the problem that I'm seeing with `exposeBin`?

Reply:
I don't really have the time right now to fiddle with this.